File: src/apps/services/src/controllers/key/key.controller.ts
import * as tedious from 'tedious';
import { Request, Response } from 'express';
import IKey from '../../models/key.interface';
import { connection as conn } from '../../main';
export default class KeyController {
  public createKey = (req: Request, res: Response) => {
    const { owner, APIKey, remainingCalls } = req.body;
    const query =
      `INSERT INTO [dbo].[keys] (owner, APIKey, remainingCalls)` +
      ` VALUES ('${owner}', '${APIKey}', ${remainingCalls})`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else {
            console.log('Express: Key created successfully.');
            return res.status(200).json({
              message: 'Key created successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public createBusinessKey = (req: Request, res: Response) => {
    const {
      owner,
      APIKey,
      remainingCalls,
      description,
      location,
      website,
      phoneNumber,
      email,
    } = req.body;
    const query =
      `INSERT INTO [dbo].[keys] (owner, APIKey, remainingCalls, isBusiness, description, location, website, phoneNumber, email)` +
      ` VALUES ('${owner}', '${APIKey}', ${remainingCalls}, 1, '${description}', '${location}', '${website}', '${phoneNumber}', '${email}')`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else {
            console.log('Express: Key created successfully.');
            return res.status(200).json({
              message: 'Key created successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public getAllKeys = (req: Request, res: Response) => {
    const query = 'SELECT * FROM [dbo].[keys]';
    const keys: IKey[] = [];

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: No keys exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'No keys exist.',
            });
          } else {
            console.log('Express: keys retrieved successfully.');
            res.status(200).json(keys);
          }
        }
      );

      request.on('row', (columns: tedious.ColumnValue[]) => {
        const key: IKey = {
          keyId: columns[0].value,
          APIKey: columns[1].value,
          remainingCalls: columns[2].value,
          suspended: columns[3].value,
          owner: columns[4].value,
          isBusiness: columns[5].value,
          description: columns[6].value,
          location: columns[7].value,
          website: columns[8].value,
          phoneNumber: columns[9].value,
          email: columns[10].value,
        };
        keys.push(key);
      });

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public getAllBusinessKeys = (req: Request, res: Response) => {
    const query = 'SELECT * FROM [dbo].[keys] WHERE isBusiness = 1';
    const keys: IKey[] = [];

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: No business keys exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'No business keys exist.',
            });
          } else {
            console.log('Express: business keys retrieved successfully.');
            res.status(200).json(keys);
          }
        }
      );

      request.on('row', (columns: tedious.ColumnValue[]) => {
        const key: IKey = {
          keyId: columns[0].value,
          APIKey: columns[1].value,
          remainingCalls: columns[2].value,
          suspended: columns[3].value,
          owner: columns[4].value,
          isBusiness: columns[5].value,
          description: columns[6].value,
          location: columns[7].value,
          website: columns[8].value,
          phoneNumber: columns[9].value,
          email: columns[10].value,
        };
        keys.push(key);
      });

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public getKey = (req: Request, res: Response) => {
    const { keyId } = req.params;
    let key: IKey;
    const query = `SELECT * FROM [dbo].[keys] WHERE keyId = ${keyId}`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: Key does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'Key does not exist.',
            });
          } else {
            console.log('Express: Key retrieved successfully.');
            res.status(200).json(key);
          }
        }
      );

      request.on('row', (columns: tedious.ColumnValue[]) => {
        key = {
          keyId: columns[0].value,
          APIKey: columns[1].value,
          remainingCalls: columns[2].value,
          suspended: columns[3].value,
          owner: columns[4].value,
          isBusiness: columns[5].value,
          description: columns[6].value,
          location: columns[7].value,
          website: columns[8].value,
          phoneNumber: columns[9].value,
          email: columns[10].value,
        };
      });

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public updateKey = (req: Request, res: Response) => {
    const { keyId } = req.params;
    const { owner, APIKey, remainingCalls, suspended } = req.body;
    const query =
      `UPDATE [dbo].[keys] SET owner = '${owner}', APIKey = '${APIKey}', remainingCalls = ${remainingCalls}, suspended = '${suspended}'` +
      ` WHERE keyId = ${keyId}`;
    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: Key does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'Key does not exist.',
            });
          } else {
            console.log('Express: Key updated successfully.');
            res.status(200).json({
              message: 'Key updated successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public updateBusinessKey = (req: Request, res: Response) => {
    const { keyId } = req.params;
    const {
      owner,
      APIKey,
      remainingCalls,
      suspended,
      description,
      location,
      website,
      phoneNumber,
      email,
    } = req.body;
    const query =
      `UPDATE [dbo].[keys] SET owner = '${owner}', APIKey = '${APIKey}', remainingCalls = ${remainingCalls}, suspended = '${suspended}',` +
      ` description = '${description}', location = '${location}' , website = '${website}', phoneNumber = '${phoneNumber}', email = '${email}'` +
      ` WHERE keyId = ${keyId}`;
    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: Key does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'Key does not exist.',
            });
          } else {
            console.log('Express: Key updated successfully.');
            res.status(200).json({
              message: 'Key updated successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public deleteKey = async (req: Request, res: Response) => {
    const { keyId } = req.params;
    const query = `DELETE FROM [dbo].[keys] WHERE keyId = ${keyId}`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: Key does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'Key does not exist.',
            });
          } else {
            console.log('Express: Key deleted successfully.');
            res.status(200).json({
              message: 'Key deleted successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };
}

File: src/apps/api/Controllers/KeyController.cs
using System.Net;
using Microsoft.AspNetCore.Mvc;
using Api.Repository;

namespace Api.Controllers;

[ApiController]
[Route("[controller]")]
public class KeyController : ControllerBase
{
    private readonly KeysRepository _keysRepository;

    public KeyController()
    {
        _keysRepository = new KeysRepository();
    }

    [HttpGet("all")]
    public async Task<IActionResult> GetAllKeys()
    {
        try
        {
            var keys = await _keysRepository.GetAllKeys();
            return Ok(keys);
        }
        catch (Exception e)
        {
            return StatusCode((int)HttpStatusCode.InternalServerError, e.Message);
        }
    }

    // all business
    [HttpGet("allBusiness")]
    public async Task<IActionResult> GetAllBusinessKeys()
    {
        try
        {
            var keys = await _keysRepository.GetAllBusinessKeys();
            return Ok(keys);
        }
        catch (Exception e)
        {
            return StatusCode((int)HttpStatusCode.InternalServerError, e.Message);
        }
    }

    //create key
    [HttpPost("create")]
    public async Task<IActionResult> CreateKey([FromBody] KeyModel keys)
    {
        try
        {
            int suspended = keys.suspended;
            var key = await _keysRepository.CreateKey(keys.owner!, keys.remainingCalls, suspended);
            return Ok(key);
        }
        catch (Exception e)
        {
            return StatusCode((int)HttpStatusCode.InternalServerError, e.Message);
        }
    }

    //create business key
    [HttpPost("createBusiness")]
    public async Task<IActionResult> CreateBusinessKey([FromBody] KeyModel keys)
    {
        try
        {
            int suspended = keys.suspended;
            var key = await _keysRepository.CreateBusinessKey(
                keys.owner!,
                keys.remainingCalls,
                suspended,
                keys.description!,
                keys.location!,
                keys.website!,
                keys.phoneNumber!,
                keys.email!
            );
            return Ok(key);
        }
        catch (Exception e)
        {
            return StatusCode((int)HttpStatusCode.InternalServerError, e.Message);
        }
    }

    //Delete a key
    [HttpDelete]
    [Route("delete/{keyId}")]
    public async Task<IActionResult> DeleteKey([FromRoute] int keyId)
    {
        try
        {
            var data = await _keysRepository.DeleteKeys(keyId);
            if (data == false)
            {
                return StatusCode(404, "Key with id: " + keyId + " does not exist");
            }
            return Ok("Deleted key with id: " + keyId + "");
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    //Update a key
    [HttpPatch]
    [Route("update")]
    public async Task<IActionResult> UpdateKey([FromBody] KeyModel key)
    {
        try
        {
            var data = await _keysRepository.UpdateKeys(
                key.keyId,
                key.owner!,
                key.APIKey!,
                key.remainingCalls,
                key.suspended
            );
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    //Update Business a key
    [HttpPatch]
    [Route("updateBusiness")]
    public async Task<IActionResult> UpdateBusinessKey([FromBody] KeyModel key)
    {
        try
        {
            var data = await _keysRepository.UpdateBusinessKeys(
                key.keyId,
                key.owner!,
                key.APIKey!,
                key.remainingCalls,
                key.suspended,
                key.isBusiness,
                key.description!,
                key.location!,
                key.website!,
                key.phoneNumber!,
                key.email!
            );
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }
}
